### ISSUE TITLE
Offset with float[] assertion doesn't have the same behavior as Offset with float assertion

### ISSUE THREAD

**Issue by juherr** [assertj/assertj-core#2414]
#### Summary
`Offset` with `float[]` assertion doesn't have the same behavior as `Offset` with `float` assertion.

Tested with `assertj-3.21.0`.

#### Example

```java
assertThat(0.2f).isEqualTo(0.7f, within(0.5f)); // pass
assertThat(new float[]{0.2f}).containsExactly(new float[]{0.7f}, within(0.5f)); // fail but should pass
```

> Expecting actual:
  [0.2f]
to contain exactly (and in same order):
  [0.7f]
but some elements were not found:
  [0.7f]
and others were not expected:
  [0.2f]
when comparing values using float comparator at precision 0.5

**Comment by scordio** [comment]
Thank you for reporting it, @juherr. We'll look into it.

**Comment by temp-droid** [comment]
The first one uses: https://github.com/assertj/assertj-core/blob/main/src/main/java/org/assertj/core/internal/Numbers.java#L191.
It checks for strict and not strict comparison.

The second one uses https://github.com/assertj/assertj-core/blob/main/src/main/java/org/assertj/core/api/ComparatorFactory.java#L45
It only checks for strict comparison. 

Should the `Offset` used in `Numbers` be strict by default?
Or should `ComparatorFactory.floatComparatorWithPrecision` take the `Offset` instead of only the value to be consistent?

I have a preference for the latter, but up to you to decide.

Also, do we need 2 separate classes for float comparisons?

**Comment by juherr** [comment]
@temp-droid In case I want strict comparisons, I would use:

```java
assertThat(0.2f).isEqualTo(0.7f, byLessThan(0.5f));
assertThat(new float[]{0.2f}).containsExactly(new float[]{0.7f}, byLessThan(0.5f));
```

**Comment by nga0315** [comment]
Hi, I would like to try working on this.

**Comment by scordio** [comment]
Thank you, @nga0315! Please give us a bit of time to review the issue and confirm the way forward. Once done, feel free to jump on it!

**Comment by YeeTone** [comment]
Hi, I'm a software testing engineering researcher. I would like to work on fixing this bug.

**Comment by YeeTone** [comment]
It should be a bug in the design of class `ComparatorFactory`. I have fixed this bug and submitted the pull request. In the new pull request, these two assertions can pass the test:
```java
assertThat(0.2f).isEqualTo(0.7f, within(0.5f)); // can pass
assertThat(new float[]{0.2f}).containsExactly(new float[]{0.7f}, within(0.5f)); // can pass as well!
```

Here are the link to the pull request, and it is my pleasure if the team can merge the bug fixing pull request.
[Pull request#2535](https://github.com/assertj/assertj-core/pull/2535)

**Comment by YeeTone** [comment]
> It should be a bug in the design of class `ComparatorFactory`. I have fixed this bug and submitted the pull request. In the new pull request, these two assertions can pass the test:
> 
> ```java
> assertThat(0.2f).isEqualTo(0.7f, within(0.5f)); // can pass
> assertThat(new float[]{0.2f}).containsExactly(new float[]{0.7f}, within(0.5f)); // can pass as well!
> ```
> 
> Here are the link to the pull request, and it is my pleasure if the team can merge the bug fixing pull request. [Pull request#2535](https://github.com/assertj/assertj-core/pull/2535)

And in order to confirm the robustness, I will submit the testcases into the pull request soon(much probable in this week).

**Comment by scordio** [comment]
Fixed by efee483f9a1be9638486f7d531a53aaa16d4b3c5.
